### ISSUE TITLE
🐛 Private class members used from constructor are incorrectly marked as unused

### ISSUE THREAD

**Issue by solita-jaakkoha2** [biomejs/biome#8138]
### Environment information

```block
CLI:
  Version:                      2.2.4
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm
  JS_RUNTIME_VERSION:           v22.18.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         unset

Biome Configuration:
  Status:                       Not set
  Path:                         unset

Workspace:
  Open Documents:               0
```

### What happened?

If a function or a class member is accessed only from the constructor, it is incorrectly marked as unused. In the following code block the `usedMember` and `usedFunction` are marked as 'unused' although they are accessed from the constructor.

```typescript
export class ClassWithMembersUsedOnlyFromConstructor {
  private usedMember: string
  private otherMember: string

  constructor() {
    this.usedMember = ""
    this.usedFunction(this.usedMember);
  }

  private usedFunction(param: string) {
    console.log(param)
  }

  public alsoUsedFunction(): string {
    return this.otherMember;
  }
}
```

https://biomejs.dev/playground/?code=ZQB4AHAAbwByAHQAIABjAGwAYQBzAHMAIABDAGwAYQBzAHMAVwBpAHQAaABNAGUAbQBiAGUAcgBzAFUAcwBlAGQATwBuAGwAeQBGAHIAbwBtAEMAbwBuAHMAdAByAHUAYwB0AG8AcgAgAHsACgAgACAAcAByAGkAdgBhAHQAZQAgAHUAcwBlAGQATQBlAG0AYgBlAHIAOgAgAHMAdAByAGkAbgBnAAoAIAAgAHAAcgBpAHYAYQB0AGUAIABvAHQAaABlAHIATQBlAG0AYgBlAHIAOgAgAHMAdAByAGkAbgBnAAoACgAgACAAYwBvAG4AcwB0AHIAdQBjAHQAbwByACgAKQAgAHsACgAgACAAIAAgAHQAaABpAHMALgB1AHMAZQBkAE0AZQBtAGIAZQByACAAPQAgACIAIgAKACAAIAAgACAAdABoAGkAcwAuAHUAcwBlAGQARgB1AG4AYwB0AGkAbwBuACgAdABoAGkAcwAuAHUAcwBlAGQATQBlAG0AYgBlAHIAKQA7AAoAIAAgAH0ACgAKACAAIABwAHIAaQB2AGEAdABlACAAdQBzAGUAZABGAHUAbgBjAHQAaQBvAG4AKABwAGEAcgBhAG0AOgAgAHMAdAByAGkAbgBnACkAIAB7AAoAIAAgACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKABwAGEAcgBhAG0AKQAKACAAIAB9AAoACgAgACAAcAB1AGIAbABpAGMAIABhAGwAcwBvAFUAcwBlAGQARgB1AG4AYwB0AGkAbwBuACgAKQA6ACAAcwB0AHIAaQBuAGcAIAB7AAoAIAAgACAAIAByAGUAdAB1AHIAbgAgAHQAaABpAHMALgBvAHQAaABlAHIATQBlAG0AYgBlAHIAOwAKACAAIAB9AAoAfQAKAA%3D%3D

### Expected result

No lint errors should be reported.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
@solita-jaakkoha2 please next time use the linter template.

What's the name of the rule that is buggy?

**Comment by ematipico** [comment]
cc @vladimir-ivanov

**Comment by ghostd** [comment]
It's `correctness/noUnusedPrivateClassMembers` (i've the regression on my project too)

**Comment by ematipico** [comment]
Probably a regression of https://github.com/biomejs/biome/pull/7684

**Comment by TimoGlastra** [comment]
We are running into the same problem after upgrading to Biome 2.3.6. It's removing valid usages of private class members.

One pattern we commonly use that seems to be removed is:

```ts
class A {
  private static b() {

 }

 public static c() {
   return A.b()
 }
}
```

Another pattern that seems to be removed:

```ts
class A {
  public c = this.b()

  private b() {

  }
}
```

But even this seems to be affected:

```ts
class A {
 private b = {}

  public c(name: string) {
    this.b[name] = 42
  }
}

```

**Comment by trino-thiagolagden** [comment]
More bug demo:
> lint/correctness/noUnusedPrivateClassMembers: This private class member is defined but never used.

https://biomejs.dev/playground/?lintRules=all&code=ZQB4AHAAbwByAHQAIABjAGwAYQBzAHMAIABTAGUAcgB2AGkAYwBlACAAewAKAAkAcwB0AGEAdABpAGMAIAAjAGkAbgBzAHQAYQBuAGMAZQA6ACAAUwBlAHIAdgBpAGMAZQA7AAoACgAJAHAAcgBpAHYAYQB0AGUAIABjAG8AbgBzAHQAcgB1AGMAdABvAHIAKAApACAAewAgAH0ACgAKAAkAcwB0AGEAdABpAGMAIABnAGUAdABJAG4AcwB0AGEAbgBjAGUAKAApADoAIABTAGUAcgB2AGkAYwBlACAAewAKAAkACQBpAGYAIAAoACEAUwBlAHIAdgBpAGMAZQAuACMAaQBuAHMAdABhAG4AYwBlACkAIAB7AAoACQAJAAkAUwBlAHIAdgBpAGMAZQAuACMAaQBuAHMAdABhAG4AYwBlACAAPQAgAG4AZQB3ACAAUwBlAHIAdgBpAGMAZQAoACkAOwAKAAkACQB9AAoACQAJAHIAZQB0AHUAcgBuACAAUwBlAHIAdgBpAGMAZQAuACMAaQBuAHMAdABhAG4AYwBlADsACgAJAH0ACgB9AAoA

**Comment by YTomm** [comment]
Hi, may i take this one?

**Comment by ematipico** [comment]
All yours @YTomm , thank you

**Comment by cpojer** [comment]
This is a major bug that turns valid code into broken code when applying fixes. Please consider reverting changes or fixing forward asap.

Here are all the situations I have found ([playground](https://biomejs.dev/playground/?lintRules=noUnusedPrivateClassMembers&code=ZQB4AHAAbwByAHQAIABjAGwAYQBzAHMAIABBAHAAcABsAGUAIAB7AAoAIAAgAHAAcgBpAHYAYQB0AGUAIABjAG8AbgBzAHQAcgB1AGMAdABvAHIAKAApACAAewAKACAAIAAgACAAdABoAGkAcwAuAGkAbgBpAHQAaQBhAGwAaQB6AGUAKAApADsACgAgACAAfQAKAAoAIAAgAHAAcgBpAHYAYQB0AGUAIABpAG4AaQB0AGkAYQBsAGkAegBlACgAKQA6ACAAdgBvAGkAZAAgAHsACgAgACAAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoACcAQQBwAHAAbABlACcAKQA7AAoAIAAgAH0ACgB9AAoACgBlAHgAcABvAHIAdAAgAGMAbABhAHMAcwAgAEIAYQBuAGEAbgBhACAAewAKACAAIABwAHIAaQB2AGEAdABlACAAYwBvAG4AcwB0AHIAdQBjAHQAbwByACgAcAByAGkAdgBhAHQAZQAgAGIAYQBuAGEAbgBhACAAPQAgAHQAcgB1AGUAKQAgAHsACgAgACAAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAHQAaABpAHMALgBiAGEAbgBhAG4AYQApADsACgAgACAAfQAKAH0ACgAKAGUAeABwAG8AcgB0ACAAYwBsAGEAcwBzACAASwBpAHcAaQAgAHsACgAgACAAcAByAGkAdgBhAHQAZQAgAHIAZQBhAGQAbwBuAGwAeQAgAGYAcgB1AGkAdAAgAD0AIAB0AHIAdQBlADsACgAgACAAcAByAGkAdgBhAHQAZQAgAGMAbwBuAHMAdAByAHUAYwB0AG8AcgAoACkAIAB7AAoAIAAgACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKABLAGkAdwBpAC4AZgByAHUAaQB0ACkAOwAKACAAIAB9AAoAfQAKAA%3D%3D)):

```tsx
export class Apple {
  private constructor() {
    this.initialize();
  }

  private initialize(): void {
    console.log('Apple');
  }
}

export class Banana {
  private constructor(private banana = true) {
    console.log(this.banana);
  }
}

export class Kiwi {
  private readonly fruit = true;
  private constructor() {
    console.log(Kiwi.fruit);
  }
}
```

After applying the fixer, it leaves this:
```tsx
export class Apple {
  private constructor() {
    this.initialize();
  }
}

export class Banana {
  private constructor(banana = true) {
    console.log(this.banana);
  }
}

export class Kiwi {
  private constructor() {
    console.log(Kiwi.fruit);
  }
}
```

**Comment by ematipico** [comment]
Can someone test this preview release and confirm if it fixes the issue?

New preview release available for `@biomejs/biome`. Install it using the following command:
 ```bash
 npm i https://pkg.pr.new/@biomejs/biome@cd5fa03
```

**Comment by xlc** [comment]
@ematipico I can confirm the new preview release fixes our case

**Comment by solita-jaakkoha2** [comment]
Same here, thank you for the quick response <3

**Comment by vladimir-ivanov** [comment]
let me know if I can help with anything. Thank you

**Comment by ematipico** [comment]
Thank you all for bearing with us. v2.3.7 is out now, and it should fix the regression. The schema file will be updated soon

**Comment by astanciu** [comment]
@ematipico The patch does not resolve the issue for me. Might be due to the `#` prefix

```ts
export class Foo {
  #name = "foo";

  constructor() {
    this.#init();
  }

  #init() {
    this.#name = "fooed";
  }

  #bar() {
    console.log("bar");
  }
}


```
becomes

```ts
export class Foo {
  constructor() {
    this.#init();
  }

  #init() {
    this.#name = "fooed";
  }
}


```

**Comment by astanciu** [comment]
It's not the `#`, this also fails:

```ts
export class Foo {
  private name = "foo";

  constructor() {
    this.init();
  }

  private init() {
    this.name = "fooed";
  }

  private bar() {
    console.log("bar");
  }
}

```
changes to 
```ts
export class Foo {
  constructor() {
    this.init();
  }

  private init() {
    this.name = "fooed";
  }
}

```

**Comment by ematipico** [comment]
@astanciu please open a new issue. The latest release fixed the OP problem. 

Plus I don't have much context regarding the rule, I only tried to do some damage control :)

**Comment by astanciu** [comment]
@ematipico I did yesterday but it was closed #8185, but I'll ask to get it reopened
